File: usertesting-job-crawler.js
// CLI usage:
// phantomjs [--ssl-protocol=any] usertesting-job-crawler.js [-v|--verbose]

var system = require('system');
var fs = require('fs');

var VERBOSE = false;
var loadInProgress = false;

// Calculate path of this file
var PWD = '';
var current_path_arr = system.args[0].split('/');
if (current_path_arr.length == 1) {
  PWD = '.';
} else {
  current_path_arr.pop();
  PWD = current_path_arr.join('/');
}

// ...from command
var configFile = PWD + '/config.json'
system.args.forEach(function(val, i) {
  if (val == '-v' || val == '--verbose') {
    VERBOSE = true;
  }
  if (val == '--config') {
    if (system.args.length == i) console.log('failed to set config - no option given');
    else {
      configFile = system.args[i + 1];
    }
  }
});

try {
  var settings = JSON.parse(fs.read(configFile));
  if (!settings.username || !settings.password || !settings.init_url) {
    console.log('Missing username, password, and/or initial URL. Exiting...');
    phantom.exit();
  }
} catch (e) {
  console.log('Could not find ' + configFile);
  phantom.exit();
}

var page = require('webpage').create();

page.onConsoleMessage = function(msg) {
  if (!VERBOSE) {
    return;
  }
  console.log(msg);
};

page.onError = function(msg, trace) {
  if (!VERBOSE) {
    return;
  }
  console.error('Error on page: ' + msg);
}

page.onCallback = function(query, msg) {
  if (query == 'username') {
    return settings.username;
  }
  if (query == 'password') {
    return settings.password;
  }
  if (query == 'report-jobs') {
    if (VERBOSE) { console.log('Found the following jobs: ' + msg); }
    else { console.log(msg); }
    return;
  }
  if (query == 'report-no-jobs') {
    if (VERBOSE) { console.log('No jobs found'); }
    else { console.log('None'); }
    return;
  }
  if (query == 'fatal-error') {
    console.log('Fatal error: ' + msg);
    phantom.exit();
  }
  return null;
}

page.onLoadStarted = function() {
  loadInProgress = true;
};
page.onLoadFinished = function() {
  loadInProgress = false;
};

page.viewportSize = {
  width: 1920,
  height: 1080
};
page.open(settings.init_url);
var steps = [
  function() { // Log in
    page.evaluate(function() {
      document.getElementById('user_email').value = window.callPhantom('username');
      document.getElementById('user_password').value = window.callPhantom('password');
      document.forms["new_user"].submit();
    });
  },
  function() { // dashboard
    page.render('ut.png');

    page.evaluate(function() {
      var tbl = document.querySelector(".available-test__table");
      var jobs = tbl.querySelector('tbody').rows;
      var jobCount = jobs.length;

      if (jobCount > 0) {
        var removals = tbl.querySelectorAll(".banner--error, .input--checkbox, .btn-group");
        for (var i = 0; i < removals.length; i++) {
          removals[i].parentNode.removeChild(removals[i]);
        }

        var html = tbl.outerHTML;
        window.callPhantom('report-jobs', html);
      }
      else {
        window.callPhantom('report-no-jobs');
      }
    });
  }
];

var i = 0;
interval = setInterval(function() {
  if (loadInProgress) {
    return;
  }
  if (!steps[i] || typeof steps[i] != "function") {
    return phantom.exit();
  }

  steps[i]();
  i++;

}, 300);

File: usertesting-job-crawler.py
#!/usr/bin/env python

# Note: for setting up email with sendmail, see: http://linuxconfig.org/configuring-gmail-as-sendmail-email-relay

import argparse
import commands
import json
import logging
import smtplib
import sys
import os

from datetime import datetime
from os import path
from subprocess import check_output
from distutils.spawn import find_executable
from email.mime.text import MIMEText
from email.mime.image import MIMEImage
from email.mime.multipart import MIMEMultipart

EMAIL_TEMPLATE = "%s"
SCREENSHOT = 'ut.png'

def notify_send_email(body, settings, use_gmail=False):
    sender = settings.get('email_from')
    recipient = settings.get('email_to', sender)  # If recipient isn't provided, send to self.

    try:
        if use_gmail:
            password = settings.get('gmail_password')
            if not password:
                logging.warning('Trying to send from gmail, but password was not provided.')
                return
            server = smtplib.SMTP('smtp.gmail.com', 587)
            server.starttls()
            server.login(sender, password)
        else:
            username = settings.get('email_username').encode('utf-8')
            password = settings.get('email_password').encode('utf-8')
            server = smtplib.SMTP(settings.get('email_server'), settings.get('email_port'))
            server.ehlo()
            server.starttls()
            server.ehlo()
            if username:
                    server.login(username, password)

        subject = "Alert: New User Testing Tests Available"
        message = body

        msg = MIMEMultipart()
        msg['Subject'] = subject
        msg['From'] = sender
        msg['To'] = ','.join(recipient)
        msg['mime-version'] = "1.0"
        msg['content-type'] = "text/html"
        msg.attach(MIMEText(message, 'html'))

        if os.path.exists(SCREENSHOT):
            fp = open(SCREENSHOT, 'rb')
            attachment = MIMEImage(fp.read())
            fp.close()
            msg.attach(attachment)

        server.sendmail(sender, recipient, msg.as_string())
        server.quit()
    except Exception:
        logging.exception('Failed to send succcess e-mail.')
        log(e)

def main(settings):
    try:
        # remove any existing screenshots
        if os.path.exists(SCREENSHOT):
            os.remove(SCREENSHOT)

        # Run the phantom JS script - output will be formatted like 'July 20, 2015'
        # script_output = check_output(['phantomjs', '%s/usertesting-job-crawler.js' % pwd]).strip()

        # determine phantomjs or fail
        cmd = find_executable('phantomjs') or find_executable('/usr/local/bin/phantomjs')
        if not cmd:
            logging.critical('Cannot find phantomjs. Make sure it is in your path.')
            return

        script_output = check_output([cmd, '--ssl-protocol=any', '%s/usertesting-job-crawler.js' % pwd, '--config', settings.get('configfile')]).strip()

        if not script_output or script_output == 'None':
            logging.info('No tests available.')
            return

    except ValueError:
        logging.critical("Couldn't convert output: {} from phantomJS script into a valid ... ".format(script_output))
        return
    except OSError:
        logging.critical("Something went wrong when trying to run usertesting-job-crawler.js. Is phantomjs is installed?")
        return

    #logging.info(msg + (' Sending email.' if not settings.get('no_email') else ' Not sending email.'))
    if not settings.get('no_email'):
        notify_send_email(script_output, settings, use_gmail=settings.get('use_gmail'))


def _check_settings(config):
    required_settings = (
        'username',
        'password'
    )

    for setting in required_settings:
        if not config.get(setting):
            raise ValueError('Missing setting %s in config.json file.' % setting)

    if config.get('no_email') == False and not config.get('email_from'): # email_to is not required; will default to email_from if not set
        raise ValueError('email_to and email_from required for sending email. (Run with --no-email or no_email=True to disable email.)')

    if config.get('use_gmail') and not config.get('gmail_password'):
        raise ValueError('gmail_password not found in config but is required when running with use_gmail option')

if __name__ == '__main__':

    # Configure Basic Logging
    logging.basicConfig(
        level=logging.DEBUG,
        format='%(levelname)s: %(asctime)s %(message)s',
        datefmt='%m/%d/%Y %I:%M:%S %p',
        stream=sys.stdout,
    )

    pwd = path.dirname(sys.argv[0])

    # Parse Arguments
    parser = argparse.ArgumentParser(description="Command line script to check for usertesting.com tests.")
    parser.add_argument('--no-email', action='store_true', dest='no_email', default=False, help='Don\'t send an e-mail when the script runs.')
    parser.add_argument('--use-gmail', action='store_true', dest='use_gmail', default=False, help='Use the gmail SMTP server instead of sendmail.')
    parser.add_argument('--config', dest='configfile', default='%s/config.json' % pwd, help='Config file to use (default is config.json)')
    arguments = vars(parser.parse_args())
    logging.info("config file is:" + arguments['configfile'])
    # Load Settings
    try:
        with open(arguments['configfile']) as json_file:
            settings = json.load(json_file)

            # merge args into settings IF they're True
            for key, val in arguments.iteritems():
                if not arguments.get(key): continue
                settings[key] = val

            settings['configfile'] = arguments['configfile']
            _check_settings(settings)
    except Exception as e:
        logging.error('Error loading settings from config.json file: %s' % e)
        sys.exit()

    # Configure File Logging
    if settings.get('logfile'):
        handler = logging.FileHandler('%s/%s' % (pwd, settings.get('logfile')))
        handler.setFormatter(logging.Formatter('%(levelname)s: %(asctime)s %(message)s'))
        handler.setLevel(logging.DEBUG)
        logging.getLogger('').addHandler(handler)

    logging.debug('Running cron with arguments: %s' % arguments)

    main(settings)
